src/main/java/com/conveyal/gtfs/PatternBuilder.java

```diff
@@ -0,0 +1,237 @@
+package com.conveyal.gtfs;
+
+import com.conveyal.gtfs.loader.BatchTracker;
+import com.conveyal.gtfs.loader.Feed;
+import com.conveyal.gtfs.loader.Requirement;
+import com.conveyal.gtfs.loader.Table;
+import com.conveyal.gtfs.model.Pattern;
+import com.conveyal.gtfs.model.PatternStop;
+import org.apache.commons.dbutils.DbUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.nio.file.Files;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Map;
+import java.util.Set;
+
+import static com.conveyal.gtfs.loader.JdbcGtfsLoader.copyFromFile;
+import static com.conveyal.gtfs.model.Entity.INT_MISSING;
+import static com.conveyal.gtfs.model.Entity.setDoubleParameter;
+import static com.conveyal.gtfs.model.Entity.setIntParameter;
+
+public class PatternBuilder {
+
+    private static final Logger LOG = LoggerFactory.getLogger(PatternBuilder.class);
+
+    private final Feed feed;
+    private static final String TEMP_FILE_NAME = "pattern_for_trips";
+
+    private final Connection connection;
+    public PatternBuilder(Feed feed) throws SQLException {
+        this.feed = feed;
+        connection = feed.getConnection();
+    }
+
+    public void create(Map<TripPatternKey, Pattern> patterns, Set<String> patternIdsLoadedFromFile) {
+        String patternsTableName = feed.getTableNameWithSchemaPrefix("patterns");
+        String tripsTableName = feed.getTableNameWithSchemaPrefix("trips");
+        String patternStopsTableName = feed.getTableNameWithSchemaPrefix("pattern_stops");
+
+        Table patternsTable = new Table(patternsTableName, Pattern.class, Requirement.EDITOR, Table.PATTERNS.fields);
+        Table patternStopsTable = new Table(patternStopsTableName, PatternStop.class, Requirement.EDITOR, Table.PATTERN_STOP.fields);
+
+        try {
+            File tempPatternForTripsTextFile = File.createTempFile(TEMP_FILE_NAME, "text");
+            LOG.info("Creating pattern and pattern stops tables.");
+            Statement statement = connection.createStatement();
+            statement.execute(String.format("alter table %s add column pattern_id varchar", tripsTableName));
```

Comment:
It's because the GTFS structure for trips doesn't include patterns which is a DT thing. The data is loaded in under the context of GTFS and then altered to work with patterns.

Reply:
Gotcha, that seems fine